modules/social_features/social_event/social_event.tokens.inc

```diff
@@ -0,0 +1,164 @@
+
+            if ($link !== NULL) {
+              $replacements[$original] = $link;
+            }
+          }
+          break;
+
+        case 'enrolled_user':
+          $id = $data['message']->getFieldValue('field_message_related_object', 'target_id');
+          if (!empty($id)) {
+            $enrollment_name = get_name_from_enrollment($id);
+
+            $replacements[$original] = $enrollment_name;
+          }
+          break;
+      }
+    }
+  }
+
+  return $replacements;
+}
+
+/**
+ * Render a link to an event as full link with title or just absolute as string.
+ *
+ * @param int $object_id
+ *   The ID of the event enrollment.
+ * @param bool $as_link
+ *   True if you want the full link otherwise just a canonical URL string.
+ *
+ * @return string|null
+ *   the Url.
+ */
+function get_link_to_event_from_enrollment($object_id, $as_link = FALSE) {
+  $entity_storage = \Drupal::entityTypeManager()
+    ->getStorage('event_enrollment');
+  /** @var \Drupal\social_event\Entity\EventEnrollment $entity */
+  $entity = $entity_storage->load($object_id);
+  if ($entity !== NULL) {
+    $event_id = $entity->getFieldValue('field_event', 'target_id');
+    $storage = \Drupal::entityTypeManager()->getStorage('node');
+    /** @var \Drupal\node\Entity\Node $event */
+    $event = $storage->load($event_id);
+
+    if ($as_link) {
+      return $event->toUrl('canonical')->toString();
+    }
+
+    return Link::fromTextAndUrl($event->getTitle(), $event->toUrl('canonical'))->toString();
+  }
+
+  return NULL;
+}
+
+/**
+ * Get the user display name from an Enrollment also for AN.
+ *
+ * @param int $id
+ *   the Enrollment entity ID.
+ *
+ * @return array|\Drupal\Component\Render\MarkupInterface|mixed|null|string
+ *   Markup for the Username.
+ *
+ * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+ * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+ */
+function get_name_from_enrollment($id) {
+  $entity_storage = \Drupal::entityTypeManager()
+    ->getStorage('event_enrollment');
+  /** @var \Drupal\social_event\Entity\EventEnrollment $entity */
+  $enrollment = $entity_storage->load($id);
+
+  // User is AN by default.
+  $enrollment_name = \Drupal::configFactory()->get('user.settings')->get('anonymous');
+
+  // If there is a Uid. Lets load the user and return his display name.
+  if ($enrollment->hasField('field_account') &&
+    $enrollment->getFieldValue('field_account', 'target_id') > 0) {
+    $entity_storage = \Drupal::entityTypeManager()
+      ->getStorage('user');
+    /** @var \Drupal\user\Entity\User $user */
+    $user = $entity_storage->load($enrollment->getFieldValue('field_account', 'target_id'));
+    return $user->getDisplayName();
+  }
+
+  // User is AN but has enrollment settings available, lets see if we can get
+  // the Firstname and lastname to show up in the notification.
+  if ($enrollment !== NULL &&
```

Comment:
Is there a possibility where AN has filled in his first name but not his last name? Then we will not be showing his first name either right? 🤔

Reply:
Correct there is no such scenario, thats why we need the field to be there, we need its values (otherwise we get nasty exceptions) so we can combine the two together.